File: akka-samples/scala-samples/src/main/scala/actors/ask/AskFromActor.scala
package actors.ask

import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorRef, ActorSystem, Behavior}
import akka.util.Timeout
import actors.ask.GreetingActor.Greet

import scala.concurrent.duration.DurationInt
import scala.util.{Failure, Success}

object AskFromActor {
  def main(args: Array[String]): Unit = {
    ActorSystem[String](guardian(), "ask-from-another-actor")
  }

  def guardian(): Behavior[String] =
    Behaviors.setup[String] { context =>
      implicit val timeout: Timeout = 3.seconds
      val g: ActorRef[Greet] = context.spawn(GreetingActor(), "dummy-greeter")

      context.ask[Greet, String](g, replyTo => Greet("Danny", replyTo)) {
        case Success(value) => value
        case Failure(_) => "Failed!"
      }

      Behaviors.receive[String] {(context, message) =>
        println(message)
        Behaviors.stopped
      }
    }
}

File: akka-samples/java-samples/src/main/java/actors/ask/AskFromActor.java
package actors.ask;

import akka.actor.typed.ActorRef;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

import java.time.Duration;

public class AskFromActor extends AbstractBehavior<String> {

    @Override
    public Receive<String> createReceive() {
        return newReceiveBuilder().onMessage(String.class, (message) -> {
            System.out.println(message);

            return Behaviors.stopped();
        }).build();
    }

    public AskFromActor(ActorContext<String> context) {
        super(context);

        ActorRef<GreetingActor.Greet> g = context.spawn(GreetingActor.create(), "dummy-greeter");

        final Duration timeout = Duration.ofSeconds(3);
        context.ask(GreetingActor.Greet.class, g, timeout, req ->
                new GreetingActor.Greet("Danny", context.getSelf()), (response, throwable) ->
                    throwable != null && response != null ? response.name : "Failed!");
    }

    public static Behavior<String> create() {
        return Behaviors.setup(AskFromActor::new);
    }

    public static void main(String[] args) {
        ActorSystem.create(AskFromActor.create(), "ask-from-another-actor");
    }
}
